posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc

```diff
@@ -0,0 +1,1428 @@
+Quarkus will do it automatically, so we can test the result of our changes right away:
+
+[source]
+----
+$ curl -X GET 'http://localhost:8080/client/search/?terms=laboratory'
+
+[
+    {
+        "assignedManager": {
+            "email": "dschrute@dundermifflin.net",
+            "id": 1,
+            "name": "Dwight Schrute",
+            "phone": "+1-202-555-0151"
+        },
+        "id": 2,
+        "name": "Aperture Science Laboratories"
+    }
+]
+----
+
+It worked: the test "laboratory" now matches the name "Aperture Science Laboratories".
+
+Analyzers are very powerful tools with tons of configuration options.
+To know more about analyzers in Elasticsearch,
+check out link:{hsearchDocUrl}#backend-elasticsearch-analysis[this section of the documentation],
+which includes a few links to lists of available analyzers, tokenizers and token filters in particular.
+
+== Indexing entity graphs
+
+Indexing an entity automatically is nice,
+but we can argue that it would have been reasonably simple to do it without Hibernate Search,
+simply by converting our entity to JSON and sending it to Elasticsearch manually,
+every time we create/update/delete a client.
+This will involve additional boilerplate code, but it can be an option.
+
+However, most of the time, we will not want to index data from just one entity, but from an entity graph.
+For example, let's assume we want to index the business manager's name as part of the client,
+so that we can search for "lapin" to easily get a list of all the clients managed by the business manager Phyllis Lapin.
+
+This is where things start getting complex:
+
+1. When the name of a business manager changes,
+we will need to load and reindex the assigned clients.
+2. When other properties of the business manager change (for example the phone number),
+we do not need to reindex the assigned clients,
+since these other properties are not indexed.
+
+These two requirements would make manually reindexing entities significantly harder to implement efficiently:
+the code would need to be aware of which of the business manager's properties are used when indexing a client,
+it would need to keep track of which properties of the business manager's are actually changed,
+and based on that would need to decide whether to load clients for reindexing or not.
+
+Add a couple more associations like this to the `Client` entity or (worse) add a few levels of nesting,
+and the simple boilerplate code will soon turn into a time sink.
+
+Fortunately, Hibernate Search handles all this transparently.
+In order to index the business manager's name as part of the client,
+only two steps are necessary.
+
+First, we will declare a field in the business manager:
+
+[source,JAVA]
+----
+@Entity
+public class BusinessManager extends PanacheEntity {
+
+	@OneToMany(mappedBy = "assignedManager")
+	public List<Client> assignedClients = new ArrayList<>();
+
+	@FullTextField(analyzer = "english") // <1>
```

Comment:
I would argue that this is not a good analyzer for names :).

Reply:
Fair enough, but I think after 1400 lines, I really should stop adding content to this post :) I'll add a comment about this near the end.